File: sdk/go/governance.go
package alpgo

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

const (
	GovernanceDir = ".governance"
	BallotsFile   = "ballots.jsonl"
)

type Vote struct {
	VoterDid   string
	BallotID   string
	Value      string
	Rationale  string
	Timestamp  string
	Signature  string
}

func NewVote(voterDid, ballotID, value, rationale, timestamp, signature string) *Vote {
	if rationale == "" {
		rationale = ""
	}
	if timestamp == "" {
		timestamp = time.Now().Format(time.RFC3339)
	}
	if signature == "" {
		signature = ""
	}
	return &Vote{
		VoterDid:  voterDid,
		BallotID:  ballotID,
		Value:     value,
		Rationale: rationale,
		Timestamp: timestamp,
		Signature: signature,
	}
}

func (v *Vote) ToDict() map[string]any {
	return map[string]any{
		"voter_did":  v.VoterDid,
		"ballot_id":  v.BallotID,
		"value":      v.Value,
		"rationale":  v.Rationale,
		"timestamp":  v.Timestamp,
		"signature":  v.Signature,
	}
}

func (v *Vote) Sign(privateKey string) string {
	// Simplified signing for Go without external crypto libs beyond stdlib
	payload := fmt.Sprintf("{\"ballot_id\":\"%s\",\"rationale\":\"%s\",\"timestamp\":\"%s\",\"value\":\"%s\",\"voter_did\":\"%s\"}",
		v.BallotID, v.Rationale, v.Timestamp, v.Value, v.VoterDid)
	v.Signature = sha256Hex(payload + privateKey)
	return v.Signature
}

type BallotRecord struct {
	BallotID    string
	PolicyID    string
	Description string
	Votes       []*Vote
	Status      string
	Quorum      int
	CreatedAt   string
	ClosedAt    string
}

func NewBallotRecord(ballotID, policyID, description string, votes []*Vote, status string, quorum int, createdAt, closedAt string) *BallotRecord {
	if status == "" {
		status = "open"
	}
	if createdAt == "" {
		createdAt = time.Now().Format(time.RFC3339)
	}
	return &BallotRecord{
		BallotID:    ballotID,
		PolicyID:    policyID,
		Description: description,
		Votes:       votes,
		Status:      status,
		Quorum:      quorum,
		CreatedAt:   createdAt,
		ClosedAt:    closedAt,
	}
}

func (b *BallotRecord) ToDict() map[string]any {
	voteDicts := []map[string]any{}
	for _, v := range b.Votes {
		voteDicts = append(voteDicts, v.ToDict())
	}
	return map[string]any{
		"ballot_id":   b.BallotID,
		"policy_id":   b.PolicyID,
		"description": b.Description,
		"votes":       voteDicts,
		"status":      b.Status,
		"quorum":      b.Quorum,
		"created_at":  b.CreatedAt,
		"closed_at":   b.ClosedAt,
	}
}

func (b *BallotRecord) Tally() map[string]int {
	counts := map[string]int{"approve": 0, "reject": 0, "abstain": 0}
	for _, v := range b.Votes {
		if _, ok := counts[v.Value]; !ok {
			counts[v.Value] = 0
		}
		counts[v.Value]++
	}
	counts["total"] = len(b.Votes)
	return counts
}

type GovernanceReport struct {
	BallotID   string
	Result     string
	Tally      map[string]int
	StartedAt  string
	FinishedAt string
}

func NewGovernanceReport(ballotID, result string, tally map[string]int, startedAt, finishedAt string) *GovernanceReport {
	if startedAt == "" {
		startedAt = time.Now().Format(time.RFC3339)
	}
	if finishedAt == "" {
		finishedAt = time.Now().Format(time.RFC3339)
	}
	return &GovernanceReport{
		BallotID:   ballotID,
		Result:     result,
		Tally:      tally,
		StartedAt:  startedAt,
		FinishedAt: finishedAt,
	}
}

func (g *GovernanceReport) ToDict() map[string]any {
	return map[string]any{
		"ballot_id":  g.BallotID,
		"result":     g.Result,
		"tally":      g.Tally,
		"started_at": g.StartedAt,
		"finished_at": g.FinishedAt,
	}
}

type PolicyBallot struct {
	AlpDir  string
	Ballots map[string]*BallotRecord
}

func NewPolicyBallot(alpDir string) *PolicyBallot {
	pb := &PolicyBallot{
		AlpDir:  alpDir,
		Ballots: make(map[string]*BallotRecord),
	}
	pb.Load()
	return pb
}

func (p *PolicyBallot) ballotsPath() string {
	return filepath.Join(p.AlpDir, GovernanceDir, BallotsFile)
}

func (p *PolicyBallot) Load() {
	path := p.ballotsPath()
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}
	// Simplified: in production, parse JSONL properly
	lines := splitLines(string(data))
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		// In production, parse JSON
		_ = line
	}
}

func (p *PolicyBallot) SaveBallot(ballot *BallotRecord) {
	d := filepath.Join(p.AlpDir, GovernanceDir)
	if err := os.MkdirAll(d, 0755); err != nil {
		return
	}
	// In production, append JSON line
	_ = d
}

func (p *PolicyBallot) OpenBallot(policyID, description string, quorum int) *BallotRecord {
	ballotID := fmt.Sprintf("ballot-%s", randomString(12))
	ballot := NewBallotRecord(ballotID, policyID, description, nil, "open", quorum, "", "")
	p.Ballots[ballotID] = ballot
	p.SaveBallot(ballot)
	return ballot
}

func (p *PolicyBallot) CastVote(ballotID, voterDid, value, rationale, privateKey string) *Vote {
	ballot := p.Ballots[ballotID]
	if ballot == nil || ballot.Status != "open" {
		return nil
	}
	vote := NewVote(voterDid, ballotID, value, rationale, time.Now().Format(time.RFC3339), "")
	if privateKey != "" {
		vote.Sign(privateKey)
	}
	ballot.Votes = append(ballot.Votes, vote)
	p.SaveBallot(ballot)
	return vote
}

func (p *PolicyBallot) CloseBallot(ballotID string) *BallotRecord {
	ballot := p.Ballots[ballotID]
	if ballot == nil || ballot.Status != "open" {
		return nil
	}
	ballot.Status = "closed"
	ballot.ClosedAt = time.Now().Format(time.RFC3339)
	p.SaveBallot(ballot)
	return ballot
}

func (p *PolicyBallot) GetBallot(ballotID string) *BallotRecord {
	return p.Ballots[ballotID]
}

func (p *PolicyBallot) ListBallots() []*BallotRecord {
	ballots := make([]*BallotRecord, 0, len(p.Ballots))
	for _, b := range p.Ballots {
		ballots = append(ballots, b)
	}
	sort.Slice(ballots, func(i, j int) bool {
		return ballots[i].CreatedAt < ballots[j].CreatedAt
	})
	return ballots
}

type GovernanceEngine struct {
	AlpDir          string
	Ballot          *PolicyBallot
	MinQuorum       int
	QualifiedVoters map[string]bool
}

func NewGovernanceEngine(alpDir string, minQuorum int) *GovernanceEngine {
	return &GovernanceEngine{
		AlpDir:          alpDir,
		Ballot:          NewPolicyBallot(alpDir),
		MinQuorum:       minQuorum,
		QualifiedVoters: make(map[string]bool),
	}
}

func NewGovernanceEngineDefault(alpDir string) *GovernanceEngine {
	return NewGovernanceEngine(alpDir, 3)
}

func (g *GovernanceEngine) Qualify(voterDid string) {
	g.QualifiedVoters[voterDid] = true
}

func (g *GovernanceEngine) Disqualify(voterDid string) {
	delete(g.QualifiedVoters, voterDid)
}

func (g *GovernanceEngine) Propose(policyID, description string, quorum *int) *BallotRecord {
	effectiveQuorum := g.MinQuorum
	if quorum != nil {
		effectiveQuorum = max(*quorum, len(g.QualifiedVoters)/2+1)
	} else {
		effectiveQuorum = max(g.MinQuorum, len(g.QualifiedVoters)/2+1)
	}
	return g.Ballot.OpenBallot(policyID, description, effectiveQuorum)
}

func (g *GovernanceEngine) Vote(ballotID, voterDid, value, rationale, privateKey string) map[string]any {
	if !g.QualifiedVoters[voterDid] {
		return map[string]any{"accepted": false, "reason": "voter_not_qualified"}
	}
	ballot := g.Ballot.GetBallot(ballotID)
	if ballot == nil || ballot.Status != "open" {
		return map[string]any{"accepted": false, "reason": "ballot_not_open"}
	}
	for _, v := range ballot.Votes {
		if v.VoterDid == voterDid {
			return map[string]any{"accepted": false, "reason": "already_voted"}
		}
	}
	vote := g.Ballot.CastVote(ballotID, voterDid, value, rationale, privateKey)
	if vote == nil {
		return map[string]any{"accepted": false, "reason": "cast_failed"}
	}
	return map[string]any{"accepted": true, "vote": vote.ToDict()}
}

func (g *GovernanceEngine) CloseAndTally(ballotID string) *GovernanceReport {
	ballot := g.Ballot.CloseBallot(ballotID)
	if ballot == nil {
		panic(fmt.Sprintf("Ballot '%s' not found or already closed.", ballotID))
	}
	return g.TallyBallot(ballot)
}

func (g *GovernanceEngine) GetReport(ballotID string) *GovernanceReport {
	ballot := g.Ballot.GetBallot(ballotID)
	if ballot == nil || ballot.Status != "closed" {
		return nil
	}
	return g.TallyBallot(ballot)
}

func (g *GovernanceEngine) ListBallots() []*BallotRecord {
	return g.Ballot.ListBallots()
}

func (g *GovernanceEngine) TallyBallot(ballot *BallotRecord) *GovernanceReport {
	tally := ballot.Tally()
	total := tally["total"]
	var result string
	if total < ballot.Quorum {
		result = "quorum_not_met"
	} else if tally["approve"] > tally["reject"] {
		result = "approved"
	} else if tally["reject"] > tally["approve"] {
		result = "rejected"
	} else {
		result = "tied"
	}
	return NewGovernanceReport(ballot.BallotID, result, tally, ballot.CreatedAt, ballot.ClosedAt)
}

func splitLines(s string) []string {
	var lines []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			lines = append(lines, s[start:i])
			start = i + 1
		}
	}
	if start < len(s) {
		lines = append(lines, s[start:])
	}
	return lines
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: sdk/rust/src/governance.rs
use std::collections::HashMap;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone)]
pub struct Vote {
    pub voter_did: String,
    pub ballot_id: String,
    pub value: String,
    pub rationale: String,
    pub timestamp: String,
    pub signature: String,
}

impl Vote {
    pub fn new(
        voter_did: impl Into<String>,
        ballot_id: impl Into<String>,
        value: impl Into<String>,
        rationale: impl Into<String>,
        timestamp: impl Into<String>,
        signature: impl Into<String>,
    ) -> Self {
        Self {
            voter_did: voter_did.into(),
            ballot_id: ballot_id.into(),
            value: value.into(),
            rationale: rationale.into(),
            timestamp: timestamp.into(),
            signature: signature.into(),
        }
    }

    pub fn to_dict(&self) -> HashMap<String, serde_json::Value> {
        let mut dict = HashMap::new();
        dict.insert(
            "voter_did".into(),
            serde_json::Value::String(self.voter_did.clone()),
        );
        dict.insert(
            "ballot_id".into(),
            serde_json::Value::String(self.ballot_id.clone()),
        );
        dict.insert(
            "value".into(),
            serde_json::Value::String(self.value.clone()),
        );
        dict.insert(
            "rationale".into(),
            serde_json::Value::String(self.rationale.clone()),
        );
        dict.insert(
            "timestamp".into(),
            serde_json::Value::String(self.timestamp.clone()),
        );
        dict.insert(
            "signature".into(),
            serde_json::Value::String(self.signature.clone()),
        );
        dict
    }

    pub fn sign(&mut self, private_key: &str) -> String {
        let mut payload = HashMap::new();
        payload.insert(
            "ballot_id",
            serde_json::Value::String(self.ballot_id.clone()),
        );
        payload.insert(
            "rationale",
            serde_json::Value::String(self.rationale.clone()),
        );
        payload.insert(
            "timestamp",
            serde_json::Value::String(self.timestamp.clone()),
        );
        payload.insert("value", serde_json::Value::String(self.value.clone()));
        payload.insert(
            "voter_did",
            serde_json::Value::String(self.voter_did.clone()),
        );
        let payload_str = serde_json::to_string(&payload).unwrap_or_default();
        self.signature = simple_hash(&(payload_str + private_key));
        self.signature.clone()
    }
}

#[derive(Debug, Clone)]
pub struct BallotRecord {
    pub ballot_id: String,
    pub policy_id: String,
    pub description: String,
    pub votes: Vec<Vote>,
    pub status: String,
    pub quorum: usize,
    pub created_at: String,
    pub closed_at: String,
}

impl BallotRecord {
    pub fn new(
        ballot_id: impl Into<String>,
        policy_id: impl Into<String>,
        description: impl Into<String>,
        votes: Vec<Vote>,
        status: impl Into<String>,
        quorum: usize,
        created_at: impl Into<String>,
        closed_at: impl Into<String>,
    ) -> Self {
        Self {
            ballot_id: ballot_id.into(),
            policy_id: policy_id.into(),
            description: description.into(),
            votes,
            status: status.into(),
            quorum,
            created_at: created_at.into(),
            closed_at: closed_at.into(),
        }
    }

    pub fn to_dict(&self) -> HashMap<String, serde_json::Value> {
        let mut dict = HashMap::new();
        dict.insert(
            "ballot_id".into(),
            serde_json::Value::String(self.ballot_id.clone()),
        );
        dict.insert(
            "policy_id".into(),
            serde_json::Value::String(self.policy_id.clone()),
        );
        dict.insert(
            "description".into(),
            serde_json::Value::String(self.description.clone()),
        );
        dict.insert(
            "votes".into(),
            serde_json::Value::Array(
                self.votes
                    .iter()
                    .map(|v| {
                        serde_json::Value::Object(
                            v.to_dict().into_iter().map(|(k, v)| (k, v)).collect(),
                        )
                    })
                    .collect(),
            ),
        );
        dict.insert(
            "status".into(),
            serde_json::Value::String(self.status.clone()),
        );
        dict.insert(
            "quorum".into(),
            serde_json::Value::Number(self.quorum.into()),
        );
        dict.insert(
            "created_at".into(),
            serde_json::Value::String(self.created_at.clone()),
        );
        dict.insert(
            "closed_at".into(),
            serde_json::Value::String(self.closed_at.clone()),
        );
        dict
    }

    pub fn tally(&self) -> HashMap<String, usize> {
        let mut counts = HashMap::new();
        counts.insert("approve".into(), 0);
        counts.insert("reject".into(), 0);
        counts.insert("abstain".into(), 0);
        for vote in &self.votes {
            let val = &vote.value;
            *counts.entry(val.clone()).or_insert(0) += 1;
        }
        counts.insert("total".into(), self.votes.len());
        counts
    }
}

pub struct GovernanceReport {
    pub ballot_id: String,
    pub result: String,
    pub tally: HashMap<String, usize>,
    pub started_at: String,
    pub finished_at: String,
}

impl GovernanceReport {
    pub fn new(
        ballot_id: impl Into<String>,
        result: impl Into<String>,
        tally: HashMap<String, usize>,
        started_at: impl Into<String>,
        finished_at: impl Into<String>,
    ) -> Self {
        Self {
            ballot_id: ballot_id.into(),
            result: result.into(),
            tally,
            started_at: started_at.into(),
            finished_at: finished_at.into(),
        }
    }

    pub fn to_dict(&self) -> HashMap<String, serde_json::Value> {
        let mut dict = HashMap::new();
        dict.insert(
            "ballot_id".into(),
            serde_json::Value::String(self.ballot_id.clone()),
        );
        dict.insert(
            "result".into(),
            serde_json::Value::String(self.result.clone()),
        );
        dict.insert(
            "tally".into(),
            serde_json::Value::Object(
                self.tally
                    .iter()
                    .map(|(k, v)| (k.clone(), serde_json::Value::Number((*v).into())))
                    .collect(),
            ),
        );
        dict.insert(
            "started_at".into(),
            serde_json::Value::String(self.started_at.clone()),
        );
        dict.insert(
            "finished_at".into(),
            serde_json::Value::String(self.finished_at.clone()),
        );
        dict
    }
}

pub struct PolicyBallot {
    alp_dir: String,
    ballots: HashMap<String, BallotRecord>,
}

impl PolicyBallot {
    pub fn new(alp_dir: impl Into<String>) -> Self {
        let alp_dir = alp_dir.into();
        let mut ballot = Self {
            alp_dir: alp_dir.clone(),
            ballots: HashMap::new(),
        };
        ballot.load();
        ballot
    }

    fn ballots_path(&self) -> String {
        Path::new(&self.alp_dir)
            .join(".governance")
            .join("ballots.jsonl")
            .to_string_lossy()
            .into_owned()
    }

    pub fn load(&mut self) {
        let path = self.ballots_path();
        if !std::path::Path::new(&path).exists() {
            return;
        }
        if let Ok(content) = std::fs::read_to_string(&path) {
            for line in content.lines() {
                if line.trim().is_empty() {
                    continue;
                }
                if let Ok(entry) = serde_json::from_str::<HashMap<String, serde_json::Value>>(line)
                {
                    let ballot_id = entry
                        .get("ballot_id")
                        .and_then(|v| v.as_str())
                        .unwrap_or("");
                    let policy_id = entry
                        .get("policy_id")
                        .and_then(|v| v.as_str())
                        .unwrap_or("");
                    let description = entry
                        .get("description")
                        .and_then(|v| v.as_str())
                        .unwrap_or("");
                    let status = entry
                        .get("status")
                        .and_then(|v| v.as_str())
                        .unwrap_or("open");
                    let quorum = entry.get("quorum").and_then(|v| v.as_u64()).unwrap_or(3) as usize;
                    let created_at = entry
                        .get("created_at")
                        .and_then(|v| v.as_str())
                        .unwrap_or("");
                    let closed_at = entry
                        .get("closed_at")
                        .and_then(|v| v.as_str())
                        .unwrap_or("");
                    let votes = Vec::new();
                    let ballot = BallotRecord::new(
                        ballot_id,
                        policy_id,
                        description,
                        votes,
                        status,
                        quorum,
                        created_at,
                        closed_at,
                    );
                    self.ballots.insert(ballot_id.into(), ballot);
                }
            }
        }
    }

    fn save_ballot(&self, ballot: &BallotRecord) {
        let dir = Path::new(&self.alp_dir).join(".governance");
        let _ = std::fs::create_dir_all(&dir);
        let path = self.ballots_path();
        if let Ok(json) = serde_json::to_string(&ballot.to_dict()) {
            let _ = std::fs::OpenOptions::new()
                .create(true)
                .append(true)
                .open(path)
                .and_then(|mut f| {
                    use std::io::Write;
                    writeln!(f, "{}", json)
                });
        }
    }

    pub fn open_ballot(
        &mut self,
        policy_id: impl Into<String>,
        description: impl Into<String>,
        quorum: usize,
    ) -> BallotRecord {
        let ballot_id = format!("ballot-{}", timestamp_now());
        let ballot = BallotRecord::new(
            &ballot_id,
            policy_id,
            description,
            Vec::new(),
            "open",
            quorum,
            timestamp_now(),
            "",
        );
        self.ballots.insert(ballot_id.clone(), ballot.clone());
        self.save_ballot(&ballot);
        ballot
    }

    pub fn cast_vote(
        &mut self,
        ballot_id: &str,
        voter_did: impl Into<String>,
        value: impl Into<String>,
        rationale: impl Into<String>,
        private_key: &str,
    ) -> Option<Vote> {
        let ballot = self.ballots.get_mut(ballot_id)?;
        if ballot.status != "open" {
            return None;
        }
        let mut vote = Vote::new(voter_did, ballot_id, value, rationale, timestamp_now(), "");
        if !private_key.is_empty() {
            vote.sign(private_key);
        }
        ballot.votes.push(vote.clone());
        let saved = ballot.clone();
        self.save_ballot(&saved);
        Some(vote)
    }

    pub fn close_ballot(&mut self, ballot_id: &str) -> Option<BallotRecord> {
        let ballot = self.ballots.get_mut(ballot_id)?;
        if ballot.status != "open" {
            return None;
        }
        ballot.status = "closed".into();
        ballot.closed_at = timestamp_now();
        let saved = ballot.clone();
        self.save_ballot(&saved);
        Some(saved)
    }

    pub fn get_ballot(&self, ballot_id: &str) -> Option<&BallotRecord> {
        self.ballots.get(ballot_id)
    }

    pub fn list_ballots(&self) -> Vec<BallotRecord> {
        self.ballots.values().cloned().collect()
    }
}

pub struct GovernanceEngine {
    alp_dir: String,
    ballot: PolicyBallot,
    min_quorum: usize,
    qualified_voters: Vec<String>,
}

impl GovernanceEngine {
    pub fn new(alp_dir: impl Into<String>, min_quorum: usize) -> Self {
        let alp_dir = alp_dir.into();
        Self {
            alp_dir: alp_dir.clone(),
            ballot: PolicyBallot::new(&alp_dir),
            min_quorum,
            qualified_voters: Vec::new(),
        }
    }

    pub fn qualify(&mut self, voter_did: impl Into<String>) {
        self.qualified_voters.push(voter_did.into());
    }

    pub fn disqualify(&mut self, voter_did: &str) {
        self.qualified_voters.retain(|v| v != voter_did);
    }

    pub fn propose(
        &mut self,
        policy_id: impl Into<String>,
        description: impl Into<String>,
        quorum: Option<usize>,
    ) -> BallotRecord {
        let effective_quorum =
            quorum.unwrap_or_else(|| self.min_quorum.max(self.qualified_voters.len() / 2 + 1));
        self.ballot
            .open_ballot(policy_id, description, effective_quorum)
    }

    pub fn vote(
        &mut self,
        ballot_id: &str,
        voter_did: impl Into<String>,
        value: impl Into<String>,
        rationale: impl Into<String>,
        private_key: &str,
    ) -> HashMap<String, serde_json::Value> {
        let voter_did = voter_did.into();
        if !self.qualified_voters.contains(&voter_did) {
            let mut result = HashMap::new();
            result.insert("accepted".into(), serde_json::Value::Bool(false));
            result.insert(
                "reason".into(),
                serde_json::Value::String("voter_not_qualified".into()),
            );
            return result;
        }
        let ballot = self.ballot.get_ballot(ballot_id);
        if ballot.is_none() || ballot.unwrap().status != "open" {
            let mut result = HashMap::new();
            result.insert("accepted".into(), serde_json::Value::Bool(false));
            result.insert(
                "reason".into(),
                serde_json::Value::String("ballot_not_open".into()),
            );
            return result;
        }
        let existing = self
            .ballot
            .get_ballot(ballot_id)
            .unwrap()
            .votes
            .iter()
            .any(|v| v.voter_did == voter_did);
        if existing {
            let mut result = HashMap::new();
            result.insert("accepted".into(), serde_json::Value::Bool(false));
            result.insert(
                "reason".into(),
                serde_json::Value::String("already_voted".into()),
            );
            return result;
        }
        let vote = self
            .ballot
            .cast_vote(ballot_id, voter_did, value, rationale, private_key);
        if vote.is_none() {
            let mut result = HashMap::new();
            result.insert("accepted".into(), serde_json::Value::Bool(false));
            result.insert(
                "reason".into(),
                serde_json::Value::String("cast_failed".into()),
            );
            return result;
        }
        let mut result = HashMap::new();
        result.insert("accepted".into(), serde_json::Value::Bool(true));
        result.insert(
            "vote".into(),
            serde_json::Value::Object(
                vote.unwrap()
                    .to_dict()
                    .into_iter()
                    .map(|(k, v)| (k, v))
                    .collect(),
            ),
        );
        result
    }

    pub fn close_and_tally(&mut self, ballot_id: &str) -> GovernanceReport {
        let ballot = self.ballot.close_ballot(ballot_id);
        if ballot.is_none() {
            panic!("Ballot '{}' not found or already closed.", ballot_id);
        }
        self.tally_ballot(&ballot.unwrap())
    }

    pub fn get_report(&self, ballot_id: &str) -> Option<GovernanceReport> {
        let ballot = self.ballot.get_ballot(ballot_id)?;
        if ballot.status != "closed" {
            return None;
        }
        Some(self.tally_ballot(ballot))
    }

    pub fn list_ballots(&self) -> Vec<BallotRecord> {
        self.ballot.list_ballots()
    }

    fn tally_ballot(&self, ballot: &BallotRecord) -> GovernanceReport {
        let tally = ballot.tally();
        let total = tally.get("total").copied().unwrap_or(0);
        let result = if total < ballot.quorum {
            "quorum_not_met"
        } else if tally.get("approve").copied().unwrap_or(0)
            > tally.get("reject").copied().unwrap_or(0)
        {
            "approved"
        } else if tally.get("reject").copied().unwrap_or(0)
            > tally.get("approve").copied().unwrap_or(0)
        {
            "rejected"
        } else {
            "tied"
        };
        GovernanceReport::new(
            ballot.ballot_id.clone(),
            result,
            tally,
            ballot.created_at.clone(),
            ballot.closed_at.clone(),
        )
    }
}

impl BallotRecord {
    pub fn ballot_id(&self) -> &str {
        &self.ballot_id
    }
}

fn timestamp_now() -> String {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    now.as_millis().to_string()
}

fn simple_hash(input: &str) -> String {
    let mut hash: u64 = 0;
    for byte in input.bytes() {
        hash = hash.wrapping_mul(31).wrapping_add(byte as u64);
    }
    format!("{:x}", hash)
}
